### PR TITLE
Support multi-column TensorFlow outputs

### PR DESCRIPTION
**Pull request by dwyatte**
## What changes are proposed in this pull request?
 
This PR addresses https://github.com/mlflow/mlflow/issues/1066, changing the behavior of the TensorFlow model flavor to not automatically ravel its output during prediction. In the case of multi-column outputs, multiple columns are returned.

This is required for TensorFlow model compatibility with Spark UDFs with multi-column outputs. I've named the outputs _output_tensor0_0_, _output_tensor0_1_, ..., _output_tensor0_n_, which doesn't make a difference in the Spark UDF case (since they it is called with `withColumn`), but may for scoring server and SageMaker/Azure functionality.
 
## How is this patch tested?

I've modified the original tests with the expected output column names. Multi-column outputs are not explicitly tested, but if this if this design looks across the other deployment options, I can add the test.
 
(Details)
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users. 

TensorFlow models with multidimensional outputs (e.g., Softmax) will not be automatically raveled during prediction.
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [ ] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [x] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [x] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

Given that this changes output shapes, I am marking as a breaking change. Presumably users handled this manually when using SageMaker/Azure for scoring and reshaped the data client-side.

### CODE REVIEW

File: mlflow/tensorflow.py [hunk 1]
```diff
@@ -344,8 +345,15 @@ def predict(self, df):
...
             raw_preds = self.tf_sess.run(self.output_tensors, feed_dict=feed_dict)
-            pred_dict = {column_name: values.ravel() for column_name, values in raw_preds.items()}
-            return pandas.DataFrame(data=pred_dict)
+            # Output tensors can be multidimensional (e.g., softmax), so return in DataFrame with
+            # columns tensor0_0, tensor0_1, ..., tensor0_n, tensor1_0, tensor1_1, ..., tensor1_n
+            pred_df = pandas.DataFrame()
+            for output_name, output_values in raw_preds.items():
+                if output_values.ndim == 1:
+                    output_values = numpy.expand_dims(output_values, axis=1)
+                for column in range(output_values.shape[1]):
```
Comment (tomasatdatabricks) [inline mlflow/tensorflow.py]: Would this work as expected with ndim > 2?

**Review by tomasatdatabricks — COMMENTED** [review]
Hi @dwyatte, thanks fro the contribution!

I think it makes sense not to ravel the outputs, but can we not return the results as multidimensional array directly?

**Comment by dwyatte** [comment]
> Hi @dwyatte, thanks fro the contribution!
> 
> I think it makes sense not to ravel the outputs, but can we not return the results as multidimensional array directly?

By returning the array as-is, you mean not instantiating a DataFrame, correct? I think this should be fine since this is the behavior of scikit-learn model outputs.

File: mlflow/tensorflow.py (hunk 1 above)
Comment (dwyatte) [reply mlflow/tensorflow.py]: I think it makes most sense to raise an exception in the ndim > 2 case, especially if MLflow's design centers largely around tabular data structures (Pandas DataFrames, Spark DataFrames). This also ensures parity across the different deployment options (Spark UDF, SageMaker/Azure, local scoring server).

Another option would be reshaping to `[N, prod(dim2, ..., dimn)]` but it may be less confusing to just have users use a custom pyfunc at that point.

Comment (tomasatdatabricks) [reply mlflow/tensorflow.py]: I don't think we should throw an exception, it would make mlflow not work with any model which produces multidimensional output.

**Comment by tomasatdatabricks** [comment]
I think pandas Dataframe can have multidiomensional arrays in them so I think we should return a pandas DataFrame with potentially multidimensional columns defined by the output tensors.

**Comment by tomasatdatabricks** [comment]
Hi @dwyatte, I have realized that this PR clashes with https://github.com/mlflow/mlflow/pull/1872. 

@juntai-zheng, can you take a look at this one to confirm?

**Comment by dwyatte** [comment]
Updated PR with the following behavior:

* Multidimensional arrays are inserted into the return DataFrame by mapping their first axis to the DataFrame's rows. This includes `[N, 1]` matrices, which will insert an array of length 1 into each row like:

| output |
|-------|
| [1] |
| [2] |
| ... | 

* Vectors (`shape=[N,]`) are inserted as one scalar per row like:

| output |
|-------|
| 1 |
| 2 |
| ... |

This seems reasonable and sufficiently flexible for Spark UDFs, although testing revealed users will need to use `mlflow.pyfunc.spark_udf` with `result_type=StringType()` and parse the results which is a bit cumbersome. It might be nice to support the ability to grab a column of ArrayType(...) directly.

I'll leave this here for folks to decide how to resolve with TensorFlow 2.0 functionality, but regardless, it would be good to have support for multi-column outputs.

**Comment by tomasatdatabricks** [comment]
Hey @dwyatte, thanks for the updates!

It makes a lot of sense. I have compared this with #1872 and I don't think there is a clash, since the #1872 changes apply only to tf >= 2.0. This PR would basically make models for tf < 2.0 behave in the same way. 

One last thing is that is technically a breaking change. I think it should be fine but I will ask for a second opinion if we need to add some logic to keep the existing behavior for models exported with earlier versions of mlflow.

**Comment by ifokeev** [comment]
Have this issue. Merge, please.

**Comment by PowerToThePeople111** [comment]
Will this be worked on in the future?
